plugins/analysis/common/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/check/Declarations.kt

```diff
@@ -58,83 +58,101 @@ internal fun <A> SolverState.checkTopLevel(
   context: ResolutionContext,
   descriptor: DeclarationDescriptor,
   declaration: Declaration,
+  isConstructor: Boolean,
   resultName: ObjectFormula,
   bodyCheck: (data: CheckData, checkPost: (finalData: CheckData) -> Unit) -> ContSeq<A>
-): ContSeq<A> {
-  // bring the constraints in (if there are any)
-  val constraints =
-    getConstraintsFor(descriptor)?.let {
-      solver.substituteDeclarationConstraints(it, mapOf(RESULT_VAR_NAME to resultName))
-    }
-  // initialize the check data
-  val initialVarInfo =
-    if (descriptor is CallableDescriptor) {
-      val thisParam =
-        (descriptor.extensionReceiverParameter ?: descriptor.dispatchReceiverParameter)?.let {
-          ParamInfo(THIS_VAR_NAME, THIS_VAR_NAME, it.type, declaration)
-        }
-      val valueParams =
-        descriptor.valueParameters.map { param ->
-          val element =
-            (declaration as? DeclarationWithBody)?.valueParameters?.firstOrNull {
-              it?.name == param.name.value
+): ContSeq<A> =
+  continuationBracket.flatMap {
```

Comment:
We should probably split this fn which is getting complex in multiple private functions for each of the steps